modules/services/nsi.js

```diff
@@ -0,0 +1,668 @@
+//   {
+//     'primary': Set(),
+//     'alternate': Set()
+//   }
+//
+function gatherKVs(tags) {
+  let primary = new Set();
+  let alternate = new Set();
+
+  Object.keys(tags).forEach(osmkey => {
+    const osmvalue = tags[osmkey];
+    if (!osmvalue) return;
+
+    const vmap = _nsi.kvt.get(osmkey);
+    if (!vmap) return;
+
+    if (osmvalue !== 'yes') {
+      primary.add(`${osmkey}/${osmvalue}`);
+    } else {
+      alternate.add(`${osmkey}/${osmvalue}`);
+    }
+  });
+
+  // Can we try a generic building fallback match? - See #6122, #7197
+  // Only try this if we do a preset match and find nothing else remarkable about that building.
+  // For example, a way with `building=yes` + `name=Westfield` may be a Westfield department store.
+  // But a way with `building=yes` + `name=Westfield` + `public_transport=station` is a train station for a town named "Westfield"
+  const preset = presetManager.matchTags(tags, 'area');
+  if (buildingPreset[preset.id]) {
+    alternate.add('building/yes');
+  }
+
+  return { primary: primary, alternate: alternate };
+}
+
+
+// `identifyTree()`
+// NSI has a concept of trees: "brands", "operators", "flags", "transit".
+// The tree determines things like which tags are namelike, and which tags hold important wikidata.
+// This takes an Object of tags and tries to identify what tree to use.
+//
+// Arguments
+//   `tags`: `Object` containing the feature's OSM tags
+// Returns
+//   `string` the name of the tree if known
+//   or 'unknown' if it could match several trees (e.g. amenity/yes)
+//   or null if no match
+//
+function identifyTree(tags) {
+  let unknown;
+  let t;
+
+  // Check all tags
+  Object.keys(tags).forEach(osmkey => {
+    if (t) return;  // found already
+
+    const osmvalue = tags[osmkey];
+    if (!osmvalue) return;
+
+    const vmap = _nsi.kvt.get(osmkey);
+    if (!vmap) return;  // this key is not in nsi
+
+    if (osmvalue === 'yes') {
+      unknown = 'unknown';
+    } else {
+      t = vmap.get(osmvalue);
+    }
+  });
+
+  return t || unknown || null;
+}
+
+
+// `gatherNames()`
+// Gather all the namelike values that we will run through the NSI matcher.
+// It will gather values primarily from tags `name`, `name:ru`, `flag:name`
+//  and fallback to alternate tags like `brand`, `brand:ru`, `alt_name`
+//
+// Arguments
+//   `tags`: `Object` containing the feature's OSM tags
+// Returns
+//   `Object` containing namelike values to test:
+//   {
+//     'primary': Set(),
+//     'fallbacks': Set()
+//   }
+//
+function gatherNames(tags) {
+  const empty = { primary: new Set(), alternate: new Set() };
+  let primary = new Set();
+  let alternate = new Set();
+  let foundSemi = false;
+  let testNameFragments = false;
+  let patterns;
+
+  // Patterns for matching OSM keys that might contain namelike values.
+  // These roughly correspond to the "trees" concept in name-suggestion-index,
+  let t = identifyTree(tags);
+  if (!t) return empty;
+
+  if (t === 'transit') {
```

Comment:
I don't really remember, but I think it was because:
-  in NSI it's easy to match what tree a feature belongs to, because the data is organized that way... 
-  in iD, it's a lot more complicated because it can only rely on a feature's tags, and those tags need to be matched in a certain order in some cases - primary before alternate + the exception for flag countries, and a feature can possibly match multiple trees.